### DISCUSSION TITLE
Loop Closure Parameters

### DISCUSSION THREAD

**Discussion by mattiasmar** [introlab/rtabmap#1078] [Q&A] [answered]
I'm reviewing pairs of images, taken from a region where I had expected a loop closure detection.
When offline performing keypoint extraction and feature matching for these pairs of images I see a significant number of matches (tens of matches).
For the same image pairs in the databaseviewer I see no keypoint matches (using the same keypoint extractor and matcher as before). I also see no red edges connecting these image pair in the graph viewer.

What could explain this discrepancy? How would you suggest to debug it?
Does RTabMap have a preference for short loop closure edges over longer ones?
This image visualizes the scenario that I'm looking at.
![image](https://github.com/introlab/rtabmap/assets/48177580/65ca9dda-3724-4f46-a293-774519a905de)
Not sure if it's a hint to my problem, but the logs says frequently:
`[ WARN] (2023-07-03 18:34:09.572) Rtabmap.cpp:1864::process() Loop detection is not performed (rehearsedId=0, signature weight=0, _proximityByTime=0, isIncremental=1)   `

**Comment by matlabbe** [comment, marked as answer]
Which commit version is rtabmap? I cannot see this log in code base?
https://github.com/introlab/rtabmap/blob/b8c298efd9cff6da579ee125e408c212de108c37/corelib/src/Rtabmap.cpp#L1864

For visual proximity detection, there is a maximum radius `RGBD/ProximityPathFilteringRadius` (default 1 meter). You can disable it by setting it to 0. If it was also not detected by global loop closure,  it means the loop closure hypothesis was not high enough. From the trajectory you taken, it is likely that it was small (it would require many frames at same location to increase faster the hypothesis).

Don't know how you did `When offline performing keypoint extraction and feature matching`, but if you want to compare with same parameters in the database, an easy way to test matching between two nodes is to use rtabmap-databaseViewer, open Graph View and Constraints View. Using two main sliders, select both nodes you want to compare, then in Constraints View, click on Add button to add a loop closure manually.
